[CIR] Support wide string literals in CIR codegen

adams381 opened the pull request:
This PR migrates support for wide string literals from the incubator to upstream.

## Changes

- Implement wide string literal support in `getConstantArrayFromStringLiteral`
- Handle wchar_t, char16_t, and char32_t string literals
- Collect code units and create constant arrays with IntAttr elements
- Use ZeroAttr for null-filled strings

## Testing

- Copied `wide-string.cpp` test file from incubator
- Expanded test to include wchar_t test cases (incubator only had char16_t and char32_t)
- All tests pass

adams381 committed:
[CIR] Support wide string literals in CIR codegen

Implement support for wide string literals (wchar_t, char16_t, char32_t)
in getConstantArrayFromStringLiteral. This migrates the feature from
the incubator to upstream.

The implementation handles wide character string literals by:
- Collecting code units using getCodeUnit()
- Creating constant arrays with IntAttr elements
- Using ZeroAttr for null-filled strings

Add test file wide-string.cpp copied from incubator, expanded with
wchar_t test cases.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -962,9 +964,36 @@ CIRGenModule::getConstantArrayFromStringLiteral(const StringLiteral *e) {
+
+  auto arrayEltTy = mlir::dyn_cast<cir::IntType>(arrayTy.getElementType());
+  assert(arrayEltTy &&
+         "string literal elements must be emitted as integral type");
```
```suggestion
  auto arrayEltTy = mlir::cast<cir::IntType>(arrayTy.getElementType());
```

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -962,9 +964,36 @@ CIRGenModule::getConstantArrayFromStringLiteral(const StringLiteral *e) {
+         "string literal elements must be emitted as integral type");
+
+  auto arraySize = arrayTy.getSize();
+  auto literalSize = e->getLength();
```
Don't use auto here.

andykaylor commented on clang/test/CIR/CodeGen/wide-string.cpp:
```diff
@@ -0,0 +1,40 @@
+// RUN: %clang_cc1 -std=c++17 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s
```
Please add LLVM and OGCG checks here.

xlauko commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -960,9 +964,36 @@ CIRGenModule::getConstantArrayFromStringLiteral(const StringLiteral *e) {
+  SmallVector<mlir::Attribute, 32> elements;
+  elements.reserve(arraySize);
+  for (uint64_t i = 0; i < arraySize; ++i)
+    elements.push_back(cir::IntAttr::get(arrayEltTy, elementValues[i]));
```
```suggestion
  for (uint32_t elementValue : elementValues)
    elements.push_back(cir::IntAttr::get(arrayEltTy, elementValue));
```

adams381 committed:
Address reviewer feedback: improve code style and performance

- Use mlir::cast<> instead of dyn_cast<> + assert
- Use explicit types (uint64_t, unsigned) instead of auto for method returns
- Remove unnecessary SmallVector size parameters
- Optimize zero-check to happen before building vector (early exit)
- Avoid double-looping by building elements directly
- Add LLVM and OGCG checks to test file

adams381 commented:
Addressed all reviewer feedback:

**Code Style Improvements:**
- Changed `dyn_cast<>` + `assert` to `cast<>` (lines 967, 969)
- Replaced `auto` with explicit types (`uint64_t`, `unsigned`) for method returns (lines 971, 972)
- Removed unnecessary SmallVector size parameters (line 989)

**Performance Optimizations:**
- Moved zero-check before building the vector to enable early exit (lines 977-983)
- Eliminated double-looping by building elements directly from string literal instead of intermediate vector
- Removed intermediate `elementValues` vector entirely

**Test Improvements:**
- Added LLVM and OGCG checks to test file (as requested)

All reviewer suggestions have been implemented. The code now follows LLVM coding standards and includes performance optimizations.

andykaylor commented on clang/test/CIR/CodeGen/wide-string.cpp:
```diff
@@ -0,0 +1,56 @@
+// LLVM: @{{.+}} = private constant [5 x i32] zeroinitializer
+// OGCG: @{{.+}} = private unnamed_addr constant [5 x i32] zeroinitializer
+
+#include <stddef.h>
```
```suggestion
typedef __WCHAR_TYPE__ wchar_t;
```

andykaylor reviewed (commented):
This looks good. I just have a few more suggestions.

adams381 committed:
[CIR] Address reviewer feedback: assert equivalence and improve tests

- Replace padding logic with assert that arraySize == literalSize
- Improve assert message clarity
- Replace #include <stddef.h> with typedef __WCHAR_TYPE__ wchar_t
- Add test case using typedef __CHAR16_TYPE__ char16_t
- Add explanatory comments for test organization

Addresses reviewer comments on PR #171541.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -962,9 +964,36 @@ CIRGenModule::getConstantArrayFromStringLiteral(const StringLiteral *e) {
+
+  uint64_t arraySize = arrayTy.getSize();
+  unsigned literalSize = e->getLength();
+  assert(arraySize == literalSize &&
```
I see that this assert is firing in testing. I wasn't expecting that. Looking closer at the classic codegen implementation, it fills the `elements` array from zero to `E->getLength()` (i.e. `literalSize` here) and then resizes the array to `NumElements`. Given that classic codegen is using a vector of integers, maybe the resize handles the zero padding. I guess you need to bring back the zero-padding loop. Sorry about the bad suggestion.

adams381 committed:
[CIR] Fix wide string literal assertion: account for null terminator

- Change assert to arraySize == literalSize + 1 to account for null terminator
- Explicitly add null terminator to elements array
- Remove typedef statements that conflict with C++17 built-in types

Fixes CI assertion failure in PR #171541.

andykaylor reviewed (approved):
This looks great.

andykaylor committed:
Merge branch 'main' into wide-string-literal-v2
